File: src/levels/L02-factory-method/ts/factory.ts
abstract class Branch {
  constructor(private barista: Barista) {}

  // El flujo es el mismo para todas; solo `createDrink` cambia.
  // region: Branch#take
  take(order: Order) {
    const drink = this.createDrink()
    this.barista.prepare(drink)
  }
  // endregion

  abstract createDrink(): object
}

class CentroBranch extends Branch {
  // region: CentroBranch#create_drink
  createDrink() { return new Latte() }
  // endregion
}

class PlayaBranch extends Branch {
  // region: PlayaBranch#create_drink
  createDrink() { return new Frappe() }
  // endregion
}

class MontanaBranch extends Branch {
  // region: MontanaBranch#create_drink
  createDrink() { return new HotChocolate() }
  // endregion
}

File: src/levels/L02-factory-method/factory.rb
class Branch
  def initialize(barista) = @barista = barista

  # El flujo es el mismo para todas; solo `create_drink` cambia.
  # region: Branch#take
  def take(order)
    drink = create_drink
    @barista.prepare(drink)
  end
  # endregion

  def create_drink = raise(NotImplementedError)
end

class CentroBranch < Branch
  # region: CentroBranch#create_drink
  def create_drink = Latte.new
  # endregion
end

class PlayaBranch < Branch
  # region: PlayaBranch#create_drink
  def create_drink = Frappe.new
  # endregion
end

class MontanaBranch < Branch
  # region: MontanaBranch#create_drink
  def create_drink = HotChocolate.new
  # endregion
end
